tlvc/arch/i86/drivers/block/idequery.c

```diff
@@ -74,7 +74,7 @@ int INITPROC get_ide_data(int drive, struct drive_infot *drive_info) {
 	word_t port = io_ports[drive >> 1];
 	int retval = 0, i, timeout = 10000;
 
-	word_t *ide_buffer = (word_t *)heap_alloc(512, 0);
+	word_t *ide_buffer = (word_t *)heap_alloc(512, HEAP_TAG_DRVR);
 
 	out_hd(drive, IDE_DRIVE_ID);
     	while (WAITING(port) && timeout--);
```

Comment:
This code was changed in ELKS to a real time 30ms timeout which will work on very fast machines:
```
#define WAIT_READY      (3*HZ/100) /* 3/100 sec = 30msec - should be instantaneous */
        unsigned long timeout = jiffies + WAIT_READY; /* 30 ms */
        out_hd(drive, IDE_DRIVE_ID);
        while ((STATUS(port) & 0x80) == 0x80) {     /* wait 30ms until not busy */
        if (time_after(jiffies, timeout))
            goto out;               /* timeout */
        }
```

Reply:
Thanks, I'll take advantage of that. Appreciated.

Reply:
Actually @ghaerr, the `goto out` might as well be a `break`, right?

Also, further down in the ELKS code, I think the `if (drive == 2) {` is probably superfluous since the routine will be called again for the 4th drive.

Reply:
@ghaerr, 
not part of a plan, but I happened to use the compaq286 with its original 40m wide drive to test the idequery changes. It turns out that 30ms wait doesn't cut it. Changed it to 300ms and we're fine. Since the loop will break as soon as the drive is ready, the time doesn't really matter, so a high number is fine.

A different question is whether the more accurate timing is meaningful in this case.

Reply:
Wow, changing the timeout by a factor of 10 seems huge. How do you know the actual time accuracy when you're using `while(timeout--);` instead of jiffies?

> A different question is whether the more accurate timing is meaningful in this case.

I would say its always better to use a jiffies timer than otherwise, if for no other reason than to avoid problems with high speed emulators. Busy loops in general are bad practice for kernels, but we get away with it since we're non-reentrant, single CPU (at least for now!).

I'll take a look at the `if (drive == 2)`, thanks.

Reply:
Well, in this case there is no need for any accuracy, it's just a matter of waiting for however long as it takes. Since the loop exits whenever the drive is ready, it is more like a watchdog timer than a specific wait. This being in the probe phase of system startup, it's not critical.